build-support/bin/generate_travis_yml.py

```diff
@@ -0,0 +1,771 @@
+
+
+def linux_fuse_shard() -> Dict:
+  return {
+    "os": "linux",
+    "dist": "xenial",
+    "sudo": "required",
+    "python": ["2.7", "3.6", "3.7"],
+    "before_install": _linux_before_install() + [
+      "sudo apt-get install -y pkg-config fuse libfuse-dev",
+      "sudo modprobe fuse",
+      "sudo chmod 666 /dev/fuse",
+      "sudo chown root:$USER /etc/fuse.conf",
+    ]
+  }
+
+
+def _osx_env_with_pyenv(python_version: PythonVersion) -> List[str]:
+  return [
+    'PATH="/usr/local/opt/openssl/bin:${PATH}"',
+    'LDFLAGS="-L/usr/local/opt/openssl/lib"',
+    'CPPFLAGS="-I/usr/local/opt/openssl/include"',
+    'PATH="${PYENV_ROOT}/versions/${PYENV_PY27_VERSION}/bin:${PATH}"',
+    f'PATH="${{PYENV_ROOT}}/versions/${{PYENV_PY{python_version.number}_VERSION}}/bin:${{PATH}}"'
+  ]
+
+
+def osx_shard(
+  *,
+  load_test_config: bool = True,
+  python_version: PythonVersion = PythonVersion.py36,
+  osx_image: Optional[str] = None,
+) -> Dict:
+  setup = {
+    "os": "osx",
+    "language": "generic",
+    "addons": {"brew": {"packages": ["openssl"]}},
+    "before_script": [
+      "ulimit -c unlimited",
+      "ulimit -n 8192",
+    ],
+    "before_install": [
+      'curl -L https://github.com/stedolan/jq/releases/download/jq-1.5/jq-osx-amd64 -o /usr/local/bin/jq',
+      'chmod 755 /usr/local/bin/jq',
+      './build-support/bin/install_aws_cli_for_ci.sh',
+      f'./build-support/bin/install_python_for_ci.sh "${{PYENV_PY27_VERSION}}" "${{PYENV_PY{python_version.number}_VERSION}}"',
+    ],
+    "env": _osx_env_with_pyenv(python_version),
+    "stage": python_version.default_stage().value
+  }
+  if osx_image is not None:
+    setup["osx_image"] = osx_image
+  if load_test_config:
+    setup["before_script"].append(AWS_GET_PANTS_PEX)
+  return setup
+
+# ----------------------------------------------------------------------
+# Bootstrap engine
+# ----------------------------------------------------------------------
+
+def _bootstrap_command(*, python_version: PythonVersion) -> List[str]:
+  # Note that for each platform, we have the Python 3.6 shard also create fs_util and upload to S3,
+  # to take advantage of the Rust code built during bootstrapping. We use the Python 3.6 shard, as
+  # it runs during both daily and nightly CI. This requires setting PREPARE_DEPLOY=1.
+  command = [f'./build-support/bin/ci.py --bootstrap --python-version {python_version.decimal}']
+  if python_version.is_py36:
+    command.append('./build-support/bin/release.sh -f')
+  return command
+
+
+def _bootstrap_env(*, python_version: PythonVersion, platform: Platform) -> List[str]:
+  env = [
+    f'CACHE_NAME=bootstrap.{platform}.py{python_version.number}',
+    f'BOOTSTRAPPED_PEX_KEY_SUFFIX=py{python_version.number}.{platform}'
+  ]
+  if python_version.is_py36:
+    env.append("PREPARE_DEPLOY=1")
+  return env
+
+
+def bootstrap_linux(python_version: PythonVersion) -> Dict:
+  command = " && ".join(_bootstrap_command(python_version=python_version))
```

Comment:
Similarly, `safe_shlex_join()` could used here. If I was writing this PR, I would try to rewrite all the command concatenations in this file to use `safe_shlex_join()`, but it's not clear how much effort that would be to do, so it doesn't need to block this PR.